pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+	// so defer closure.
+	defer func() {
+		err = errors.Join(err, unix.Close(fd))
+	}()
+
+	localHandle := NetNSHandle{
+		fd: int(fd),
+	}
+	err = localHandle.Do(runInside)
+	return err
+}
+
+// FD returns the underlying file descriptor of the NetNSHandle
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNSHandle) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all references to it
+// (including the one being closed here) are gone.
+func (h *NetNSHandle) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
+// will be locked with `runtime.LockOSThread()` during this operation.
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the callback in a new green thread so that if we later fail
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	var innerError error
+	go func() {
+		// save a handle to current network namespace
+		// and lock the thread - this will force this goroutine to be
+		// scheduled in a different thread, so we should be safe to mutate the netns
+		restore, err := saveRestoreCurrentThreadNetns()
+		if err != nil {
+			innerError = err
+			return
+		}
+		defer restore()
+		defer wg.Done()
+		unix.Setns(h.fd, unix.CLONE_NEWNET)
+		innerError = runInside(h)
+	}()
+	wg.Wait()
+
+	return innerError
+}
+
+// NOTE If we ever need to test equality of netNSes outside of test code,
+// make this public.
+func (h *NetNSHandle) equal(other *NetNSHandle) bool {
+	if h.FD() == other.FD() {
+		return true
+	}
+
+	// Just because the FDs are not equal doesn't mean they don't both refer
+	// to the same underlying netns, so check inode too.
+	var s1, s2 unix.Stat_t
+	if err := unix.Fstat(int(h.FD()), &s1); err != nil {
+		return false
+	}
+	if err := unix.Fstat(int(other.FD()), &s2); err != nil {
+		return false
+	}
+	return (s1.Dev == s2.Dev) && (s1.Ino == s2.Ino)
+}
+
+// saveRestoreCurrentThreadNetns returns a cleanup func (which callers should `defer`)
+// and an error if for some reason the current netns is unobtainable.
+//
+// the cleanup func will move the thread back to the netns it was running under when
+// this function was called, and unlock the current OS thread.
+func saveRestoreCurrentThreadNetns() (func(), error) {
```

Comment:
The closure returned by this function attempts to undo the effects of `Unshare()`, but the parent function only calls LockOSThread(). That doesn't make any sense.

I'd either name this `lockAndEnter(fd int) (func(), error)` and make this function enter the netns after locking the OS thread, or call it `lock() (func(), error)` and have it only deal with `runtime.(Unlock)OSThread()` and a separate `enter(fd int) (func(), error)` that deals with `Unshare/Setns`.

Reply:
Yep, it's too tightly coupled to being called in concert with other helper funcs, fair enough.

Reply:
> The closure returned by this function attempts to undo the effects of Unshare(), but the parent function only calls LockOSThread(). That doesn't make any sense.

It's designed to unconditionally restore the current thread to whatever netns it had when you first invoked `saveRestoreCurrentThreadNetns`. It doesn't care what the netns was when you called that.

Reply:
> all it lock() (func(), error) and have it only deal with runtime.(Unlock)OSThread() and a separate enter(fd int) (func(), error) that deals with Unshare/Setns.

The latter would always need to be fenced with the former to be deterministic, so it doesn't make much sense here to separate them.

Reply:
> To me, what makes the most sense is a lockAndEnter() that locks the goroutine and unshares, then returns a function that sets the thread back to orig and unlocks the goroutine.


What's here does all of that, except the unshare-ing/moving namespaces is factored out of the save/restore function, because it's not strictly relevant to save/restore and might be done differently depending on whether we are `unshare`-ing, or opening an existing netns pin path (as we still must, at least, still, for CNI).

Could pass the pin path to `saveRestoreCurrentThreadNS` but that makes the private helper func a little more complex and less generalized, and doesn't change the outward-facing behavior at all.